examples/energy-management-app/energy-management-common/energy-evse/src/EnergyEvseDelegateImpl.cpp

```diff
@@ -148,54 +264,70 @@ Status EnergyEvseDelegate::EnableDischarging(const DataModel::Nullable<uint32_t>
  */
 Status EnergyEvseDelegate::ScheduleCheckOnEnabledTimeout()
 {
-
-    uint32_t matterEpoch = 0;
+    // Determine the relevant timeout based on current supply state
     DataModel::Nullable<uint32_t> enabledUntilTime;
-
-    if (mSupplyState == SupplyStateEnum::kChargingEnabled)
+    switch (mSupplyState)
     {
+    case SupplyStateEnum::kChargingEnabled:
         enabledUntilTime = GetChargingEnabledUntil();
+        break;
+    case SupplyStateEnum::kDischargingEnabled:
+        enabledUntilTime = GetDischargingEnabledUntil();
+        break;
+    case SupplyStateEnum::kEnabled:
+        // For enabled state, use the earliest of charging or discharging timeout
+        enabledUntilTime = getEarliestTime(GetChargingEnabledUntil(), GetDischargingEnabledUntil());
+        break;
+    default:
+        // In all other states the EVSE is disabled, no timer needed
+        return Status::Success;
     }
-    else if (mSupplyState == SupplyStateEnum::kDischargingEnabled)
+
+    if (enabledUntilTime.IsNull())
     {
-        enabledUntilTime = GetDischargingEnabledUntil();
+        ChipLogDetail(AppServer, "EVSE is enabled indefinitely, no timer needed");
+        return Status::Success;
     }
-    else
+
+    uint32_t matterEpoch = 0;
+    CHIP_ERROR err       = System::Clock::GetClock_MatterEpochS(matterEpoch);
+    if (err == CHIP_ERROR_REAL_TIME_NOT_SYNCED)
     {
-        // In all other states the EVSE is disabled
+        // Real time isn't sync'd - check again in 30 seconds
+        DeviceLayer::SystemLayer().StartTimer(System::Clock::Seconds32(kPeriodicCheckIntervalRealTimeClockNotSynced_sec),
+                                              EvseCheckTimerExpiry, this);
         return Status::Success;
     }
 
-    if (enabledUntilTime.IsNull())
+    if (err != CHIP_NO_ERROR)
+    {
+        return Status::Failure; // Can't get time, skip scheduling
+    }
+
+    int32_t delta = static_cast<int32_t>(enabledUntilTime.Value() - matterEpoch);
+    if (delta > 0)
```

Comment:
Might be better in terms of avoiding overflow to do:
```suggestion
    if (enabledUntilTime.Value() > matterEpoch)
```